solutions/usaco-994.mdx

```diff
@@ -2,72 +2,110 @@
+
+To do this, we can precalculate the number of triplets which start at $i$ and end at $j$ in $\mathcal{O}(N^2)$ time by storing the occurences of elements between $i$ and $j$ in an array.
+
+More formally, our final transition is:
+
+$\texttt{ways}[i][j] = \texttt{ways}[i + 1][j] + \texttt{ways}[i][j - 1] - \texttt{ways}[i + 1][j - 1] + \texttt{trp}[i][j]$
+
+where $\texttt{trp}[i][k]$ equals the number of triplets which sum to zero and start at $i$ and end at $k$.
+
+Note that we can't store $\texttt{trp}$ in it's own array due to memory constraints.
+
+
 ## Implementation
 
 **Time Complexity:** $\mathcal{O}(N^2)$
 
 <LanguageSection>
-<JavaSection>
-
-```java
-import java.util.*;
-import java.io.*;
-
-public class Threesum {
-	static final int MAX_VAL = 1000000;
-	public static void main(String[] args) throws IOException {
-		Kattio io = new Kattio("threesum");
-
-		int N = io.nextInt();
-		int Q = io.nextInt();
-		int[] a = new int[N];
-		for (int i = 0; i < N; i++) {
-			a[i] = io.nextInt();
-		}
 
-		long[][] dp = new long[N + 1][N + 1];
-		int[] cnt = new int[2 * MAX_VAL + 1];
+<CPPSection>
 
-		// Calculate the number of k such that a[i] + a[j] + a[k] = 0 given i and j
-		for (int i = 0; i < N - 1; i++) {
-			for (int j = 0; j < N; j++) {
-				cnt[a[j] + MAX_VAL] = 0;
-			}
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
 
-			for (int j = i+1; j < N; j++) {
-				int k = -a[i] - a[j];
-				if(k >= -MAX_VALUE && k <= MAX_VALUE) {
-					dp[i + 1][j + 1] += cnt[k + MAX_VAL];
-				}
+using ll = long long;
+const int max_val = 1e6;
 
-				cnt[a[j] + MAX_VAL]++;
-			}
-		}
+int main() {
+	freopen("threesum.in", "r", stdin);
+	freopen("threesum.out", "w", stdout);
+	cin.tie(0)->sync_with_stdio(0);
+	
+	int n, q;
+	cin >> n >> q;
+
+	vector<int> val(n);
+	for (int i = 0; i < n; i++) {
+		cin >> val[i];
+		val[i] += max_val;
+	}
+
+	// number of triplets such that i and k are fixed
+	vector<vector<ll>> ways(n, vector<ll>(n, 0));
+	vector<int> frq(2 * max_val);
+
+	for (int i = n - 2; i >= 0; i--) {
+		/*
+		 * be careful to only consider values
+		 * in between i and j
+		 */
+		frq[val[i + 1]]++;
 		
-		for (int i = N; i >= 1; i--) {
-			for (int j = i + 1; j <= N; j++) {
-				dp[i][j] += dp[i + 1][j] + dp[i][j - 1] - dp[i + 1][j - 1];
+		for (int j = i + 2; j < n; j++) {
+			// separate val from shifted amt
+			int rem = (max_val * 3) - (val[i] + val[j]);
+			
+			if (rem >= 0 && rem < (2 * max_val)) {
+				ways[i][j] = frq[rem];
 			}
+			frq[val[j]]++;
 		}
 
-		for (int i = 0; i < Q; i++) {
-			int a1 = io.nextInt();
-			int b = io.nextInt();
-			io.println(dp[a1][b]);
+		for (int j = i + 1; j < n; j++) {
+			frq[val[j]]--;
 		}
+	}
 
-		io.close();
+	// essentially 2D prefix sum on ways
+	for (int i = n - 1; i >= 0; i--) {
+		for (int j = i + 1; j < n; j++) {
+			ways[i][j] += (
+				ways[i + 1][j] +
+				ways[i][j - 1] -
+				ways[i + 1][j - 1]
+			);
+		}
 	}
-	
-	//CodeSnip{Kattio}
-}
 
+	for (int _ = 0; _ < q; _++) {
```

Comment:
in c++ i don't think we use underscores

Reply:
wwhat, is it not good practice?

Reply:
idk, just haven't seen anyone use it
you can keep it i guess